src/monodroid/jni/monodroid-glue.cc

```diff
@@ -1151,27 +1151,91 @@ MonodroidRuntime::monodroid_dlopen (const char *name, int flags, char **err, [[m
 {
 	unsigned int dl_flags = monodroidRuntime.convert_dl_flags (flags);
 	bool libmonodroid_fallback = false;
-
+	bool name_is_full_path = false;
+	bool name_needs_free = false;
 	/* name is nullptr when we're P/Invoking __Internal, so remap to libxa-internal-api */
-	if (name == nullptr) {
+	if (name == nullptr || strstr (name, "xa-internal-api") != nullptr) {
+#if defined (WINDOWS)
+		char *tmp_name = nullptr;
+
+		auto probe_dll_at = [&](const char *the_path) -> bool {
+			const char *last_sep = strrchr (the_path, MONODROID_PATH_SEPARATOR_CHAR);
+			if (last_sep != nullptr) {
+				tmp_name = utils.strdup_new (the_path, last_sep - the_path);
+				tmp_name = utils.string_concat (tmp_name, MONODROID_PATH_SEPARATOR, API_DSO_NAME);
+				if (!utils.file_exists (tmp_name)) {
+					delete[] tmp_name;
+					tmp_name = nullptr;
+					return false;
+				}
+
+				return true;
+			}
+
+			return false;
+		};
+
+		// First try to see if it exist at the path pointed to by `name`. With p/invokes, currently (Sep 2020), we can't
+		// really trust the path since it consists of *some* directory path + p/invoke library name and it does not
+		// point to the location where the native library is. However, we still need to check the location first, should
+		// it point to the right place in the future.
+		bool found = probe_dll_at (name);
+		if (!found) {
+			// Next lets try the location of the XA runtime DLL, libxa-internal-api.dll should be next to it.
+			HMODULE hm;
+			const void *func = reinterpret_cast<const void*>(&Java_mono_android_Runtime_initInternal);
+			if (GetModuleHandleEx (GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS | GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT, static_cast<LPCSTR>(func), &hm) == 0) {
```

Comment:
This looks like `MonodroidRuntime::get_my_location()`? Perhaps that should be used instead?

Reply:
heh, forgot that we had that one :)